### PR TITLE
feat(swagger-editor): customizable codegen URLs

### LINKED ISSUE #1488: Intranet only swagger editor via docker
| Q                                   | A
| ----------------------------------- | -------
| Bug or feature request?             |  feature
| Which Swagger/OpenAPI version?      | latest
| Which Swagger-Editor version?       | latest
| How did you install Swagger-Editor? |  docker
| Which broswer & version?            | any
| Which operating system?             |  any

### Expected Behavior
Using the docker image I should be able to start up a swagger editor instance that can be served from local non internet based resources only.

### Current Behavior
Using the (latest) docker image to start up a swagger editor instance I am greeted by an editor that requests a large amount of external internet based resources without (see swagger-editor/Dockerfile or swagger-editor/docker-run.sh) any way of telling the swagger editor to serve these resources from either the docker image itself or an additional resource like the swagger-codegen docker image. The Swagger UI can use the swagger-codegen docker image and some additional config values, though.

### Possible Solution
Add feature parity between swagger ui and swagger editor docker images concerning environment variables.

### Context
At company I am working we are currently trying to centrally host internal API documentations and would very much like to keep all related services such as the Swagger UI and Editor within the intranet only. This includes browser requests to fetch additional content such as stylesheets, JavaScript files, fonts, etc. when accessing these resources. The reasons for that should be fairly obvious. Right now there appears to be no point in using (the provided) docker images as they simply (out of the box and in case of the editor) rely on internet/third-party resources anyway. Maybe I am simply being stupid and am overlooking something very, very simple but after searching through all related issues, documentations, and related search results I could get my hands on I am stumped and can't find a way around this issue. Help would be very much appreciated.

### PR DESCRIPTION
**Pull request by antoineauger**
### Description
This PR aims at allowing users to customize the URLs used by the Swagger Editor codegen features (e.g., Generate Server and Generate Client).

The URLs that can now be customized are:
- swagger2GeneratorUrl: (https://generator.swagger.io/api/swagger.json by default)
- oas3GeneratorUrl: (https://generator3.swagger.io/openapi.json by default)
 - swagger2ConverterUrl: (https://converter.swagger.io/api/convert by default)

For Docker users, this is made possible by the introduction of 3 new environment variables: `URL_SWAGGER2_GENERATOR` `URL_OAS3_GENERATOR` and `URL_SWAGGER2_CONVERTER`, respectively.

For users who create a `SwaggerEditor` object, they can now provide an additional dict object to override these URLs (see `index.html` for an example)
```js
function SwaggerEditor(options, swaggerEditorConfigUrls = null) {
  // ...
  return SwaggerUI(mergedOptions)
}
```

### Motivation and Context

This PR can be interesting for:
- users already have their own local Swagger generator server (https://github.com/swagger-api/swagger-codegen#public-pre-built-docker-images);
- users don't want to enable the codegen features (for instance for security reasons).

Fixes #1488 #2127

/cc @fgreinacher @dlouzan

### How Has This Been Tested?

For Docker env variables, created several docker containers with various values for env variables `URL_SWAGGER2_GENERATOR`, `URL_OAS3_GENERATOR` and `URL_SWAGGER2_CONVERTER`. Then logging on startup:
```js
console.log(mergedOptions)
```

Also tested that setting these variables to `null` was disabling the Swagger Editor codegen features:
```
docker run -e URL_SWAGGER2_GENERATOR="null" -e URL_OAS3_GENERATOR="null" -e URL_SWAGGER2_CONVERTER="null" -d -p 9999:8080 swagger-editor
```

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [x] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [ ] My changes do not require a change to the project documentation.
- [x] My changes require a change to the project documentation.
- [x] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [x] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by char0n** [comment]
@antoineauger thanks a lot for contributing. We'll look into the PR ASAP.

File: src/index.js [hunk 1]
```diff
@@ -65,9 +65,15 @@ const defaults = {
   swagger2ConverterUrl: "https://converter.swagger.io/api/convert",
 }
 
-export default function SwaggerEditor(options) {
+export default function SwaggerEditor(options, swaggerEditorConfigUrls = null) {
```
Comment (char0n) [inline src/index.js:68]: What would be the reason for extending the `SwaggerEditor` function with another parameter called `swaggerEditorConfigUrls`?

`SwaggerEditor` function already supports following configuration options via `options` parameter ([see code here](https://github.com/swagger-api/swagger-editor/blob/2581d92dcf984bea0b6c19d96894f5f24dd1f03a/src/index.js#L63)):

```js
  swagger2GeneratorUrl: "https://generator.swagger.io/api/swagger.json",
  oas3GeneratorUrl: "https://generator3.swagger.io/openapi.json",
  swagger2ConverterUrl: "https://converter.swagger.io/api/convert",
```

So IMHO in order to support new ENV variables we don't need to make any changes to the code.

File: index.html [hunk 2]
```diff
@@ -57,6 +57,11 @@
         SwaggerEditorStandalonePreset
       ],
       queryConfigEnabled: false,
+    },
+    {
```
Comment (char0n) [inline index.html:61]: Correct me if I'm wrong, but this looks line unnecessary change. The configuration options are already present and have their default values set:

https://github.com/swagger-api/swagger-editor/blob/2581d92dcf984bea0b6c19d96894f5f24dd1f03a/src/index.js#L63

File: README.md [hunk 3]
```diff
@@ -129,6 +129,16 @@ docker run -d -p 80:8080 -v $(pwd):/tmp -e SWAGGER_FILE=/tmp/swagger.json swagge
 docker run -d -p 80:8080 -e BASE_URL=/swagger-editor swaggerapi/swagger-editor
 ```
 
+You can also customize the different endpoints used by the Swagger Editor with the following environment variables. For instance, this can be useful if you have your own Swagger generator server:
+
+Environment variable | Default value
+--- | --- | ---
```
Comment (char0n) [inline README.md:135]: This unfortunately doesn't generate the table. There are 3 columns in table header, but only 2 columns in table body.

File: README.md [hunk 4]
```diff
@@ -129,6 +129,16 @@ docker run -d -p 80:8080 -v $(pwd):/tmp -e SWAGGER_FILE=/tmp/swagger.json swagge
...
+`URL_SWAGGER2_GENERATOR` | https://generator.swagger.io/api/swagger.json
+`URL_OAS3_GENERATOR` | https://generator3.swagger.io/openapi.json
+`URL_SWAGGER2_CONVERTER` | https://converter.swagger.io/api/convert
+
+If you simply want to run the Swagger Editor locally without the Codegen features (Generate Server and Generate Client) you can simply set the above environment variables to `null`.
```
Comment (char0n) [inline README.md:140]:
```suggestion
If you want to run the Swagger Editor locally without the Codegen features (Generate Server and Generate Client) you can set the above environment variables to `null` (`URL_SWAGGER2_CONVERTER=null`).
```

File: docker-run.sh [hunk 5]
```diff
@@ -40,6 +40,28 @@ if [[ "${QUERY_CONFIG_ENABLED}" = "true" ]]; then
   sed -i "s|queryConfigEnabled: false|queryConfigEnabled: true|" $INDEX_FILE
 fi
 
+if [[ "${URL_SWAGGER2_GENERATOR}" ]]; then
+  if [[ "$URL_SWAGGER2_GENERATOR" != "null" ]]; then
+    sed -i "s|swagger2GeneratorUrl: 'https://generator.swagger.io/api/swagger.json'|swagger2GeneratorUrl: '${URL_SWAGGER2_GENERATOR}'|g" $INDEX_FILE
```
Comment (char0n) [inline docker-run.sh:45]: To add the `swagger2GeneratorUrl` as option to `SwaggerEditorBundle` function, following shell code needs to be introduced:

```suggestion
    sed -i "s|SwaggerEditorBundle({|SwaggerEditorBundle({\n      swagger2GeneratorUrl: '${URL_SWAGGER2_GENERATOR}',|g" index.html
```

This independently injects new options to `SwaggerEditorBundle` function and should be used through out the changes in `docker-run.sh` that this PR introduced.

**Review by char0n — CHANGES_REQUESTED** [review]
Thanks for contributing! Please have a look at my code review comments.

File: src/index.js (hunk 1 above)
Comment (antoineauger) [reply src/index.js]: Also updated the PR description.

File: index.html (hunk 2 above)
Comment (antoineauger) [reply index.html]: You're right, see my answer: https://github.com/swagger-api/swagger-editor/pull/2910#discussion_r821852314

File: docker-run.sh (hunk 5 above)
Comment (antoineauger) [reply docker-run.sh]: You're right, see my answer: https://github.com/swagger-api/swagger-editor/pull/2910#discussion_r821852314

**Comment by antoineauger** [comment]
> Thanks for contributing! Please have a look at my code review comments.

Thanks for the review @char0n, this PR is way much simpler now.

I think I addressed all your comments, could you have another look at it? 🙏🏻

**Review by char0n — APPROVED** [review]
Nice work! Changes look good! I'll have to some local testing with changes from this PR before the merge.

**Comment by char0n** [comment]
Testing went fine. All works as expected. THanks again!
